File: frontend/src/models/message.ts
export type Emote = { name: string; url: string; zeroWidth: boolean }
export type Fragment =
  | ({ type: "emote" } & Emote)
  | { type: "emote-stack"; emotes: Emote[] }
  | { type: "text"; text: string }
export type Message =
  | { type: "system"; text: string }
  | {
      type: "message"
      time: Date
      fragments: Fragment[]
      username: string
      colour: number
      replying: null | MessageId
      id: MessageId
      filtered: boolean
      roles: Promise<{ id: number; name: string }[]>
    }
export type MessageId = string & { readonly __brand: unique symbol }

File: backend/src/inter/models/message.py
from typing import Literal, TypedDict

from inter.models.user import User


class TextFragment(TypedDict):
    type: Literal["text"]
    text: str


class EmoteFragment(TypedDict):
    type: Literal["emote"]
    url: str
    name: str


class Message(TypedDict):
    time: str
    message: list["Fragment"]
    user: "User"


type Fragment = TextFragment | EmoteFragment
